File: blind75/binary_search/go/find_min_in_rotated_sorted_array.go
package main

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
func findMin(nums []int) int {
	if len(nums) == 0 {
		return -1
	}
	start, end := 0, len(nums)-1
	for start+1 < end {
		mid := (start + end) / 2
		if nums[mid] > nums[end] {
			start = mid
		} else {
			end = mid
		}
	}
	return min(nums[start], nums[end])

}

// func main() {
// 	arr := []int{3, 4, 5, 1, 2}
// 	fmt.Println(findMin(arr))
// 	arr = []int{4, 5, 6, 7, 0, 1, 2}
// 	fmt.Println(findMin(arr))
// 	arr = []int{11, 13, 15, 17}
// 	fmt.Println(findMin(arr))
// }

File: blind75/binary_search/find_min_in_rotated_sorted_array.js
/**
 * @param {number[]} nums
 * @return {number}
 */
 var findMin = function (nums) {
    if (!nums.length) {
        return -1;
    }
    let start = 0, end = nums.length - 1;
    while (start + 1 < end) {
        let mid = Math.floor((start + end) / 2);
        if (nums[mid] > nums[end]) {
            start = mid;
        } else {
            end = mid;
        }
    }
    return Math.min(nums[start], nums[end]);
};